config/config.go

```diff
@@ -100,10 +100,11 @@ var netDefaultPort = map[NetProtocol]uint16{
 
 // Upstream is the definition of external DNS server
 type Upstream struct {
-	Net  NetProtocol
-	Host string
-	Port uint16
-	Path string
+	Net        NetProtocol
+	Host       string
+	Port       uint16
+	Path       string
+	CommonName *string
```

Comment:
I think using a plain string (without pointer) and empty string to represent none would be simpler.
Just make sure to add a comment to mention it's optional and can be empty.

Reply:
Yeah, I waffled back and forth about this. I've made the change.